File: TwoPointer/75.sort-colors.go
package leetcode

// @leet start
func sortColors(nums []int) {
	l, r := 0, len(nums)-1
	for i := 0; i < len(nums); {
		v := nums[i]
		if r > i && v == 2 {
			nums[i], nums[r] = nums[r], nums[i]
			r--
			continue
		}
		if l < i && v == 0 {
			nums[i], nums[l] = nums[l], nums[i]
			l++
			continue
		}
		i++
	}
}

// @leet end

File: TwoPointer/75.sort-colors.py
from typing import List


# @leet start
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        l, r = 0, len(nums) - 1
        i = 0
        while i < len(nums):
            v = nums[i]
            if v == 0 and l < i:
                nums[i], nums[l] = nums[l], nums[i]
                l += 1
                continue
            if v == 2 and i < r:
                nums[i], nums[r] = nums[r], nums[i]
                r -= 1
                continue
            i += 1


# @leet end

